src/Scheduler/GenericScheduler.php

```diff
@@ -0,0 +1,118 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tempest\Console\Scheduler;
+
+use DatePeriod;
+use DateTime;
+use DateInterval;
+use DateTimeImmutable;
+use Tempest\Console\ConsoleCommand;
+use Tempest\Console\ConsoleConfig;
+use Tempest\Console\ConsoleOutput;
+
+final class GenericScheduler implements Scheduler
+{
+    private string $path;
+    private DateTime $dayStart;
+    private DateTime $now;
+
+    public function __construct(
+        private ConsoleConfig $config,
+        private ConsoleOutput $output,
+    ) {
+        $this->path = "php tempest"; // todo: discover path
+    }
+
+    public function run(?DateTime $date = null): void
+    {
+        $now = $date ?? (new DateTime());
+
+        $this->dayStart = (clone $now)->setTime(0, 0, 0);
+        $this->now = ($now)->setTime(
+            (int) $now->format('H'),
+            (int) $now->format('i'),
+        );
+
+        $commands = $this->getCommandsToRun();
+
+        foreach ($commands as $command) {
+            $this->execute($command);
+        }
+    }
+
+    private function execute(ConsoleCommand $commandDefinition): void
+    {
+        $name = $commandDefinition->getName();
+        $this->output->writeln("Running command: {$name}");
+
+        $command = $this->buildCommand($commandDefinition);
+
+        exec($command);
+
+        $this->output->writeln("Command finished: {$name}");
+    }
+
+    private function getCommandsToRun(): array
```

Comment:
I understand you went with a stateless approach. Basically checking if the interval matches the current time. 

I want to pitch another stateful approach, that has a couple of benefits. Would love your input.

Let's say we keep track of when a command has last run. That means that determining whether a command should run or not is as simple as checking whether the interval is smaller than the difference between now and the last run.

The benefit here is that if the scheduler fails to run for an amount of time, it'll be able to start up again and finish tasks that should have been run while it was down. 

The downside is that you loose fine-grained control over _when_ a command should run. If you want it to run exactly every 5th minute after the hour, you're out of luck. 

We could of course provide _both_ options, although I would start with the stateful approach, since it significantly reduces complexity in determining whether a scheduled command should run or not. 

wdyt?

Reply:
That makes complete sense!

How would you want to handle the saving of last command run? Do we want to create a lockfile, handle that in the DB or through cache?

Reply:
> Do we want to create a lockfile, handle that in the DB or through cache?

That's the difficult question 😅 Database is out, since this package should work standalone without a DB. So either cache or a lock file. I think a lock file is better since we don't want the schedule to be lost on cache clear.

For now, I'd use a simple file like we do with discovery caching, we can always expand on it later and make it configurable, but that's out of scope for this PR

Reply:
Makes sense! Will adjust it later today

Reply:
Hey @brendt, reworked this a bit, still lacks tests and it doesn't handle sub-minute runs yet - is that something that we'd want to support from the get go or can we add it later?

Lemme know your thoughts!

Reply:
Nah we don't need sub-minute runs atm 👍